# Total Laps and Average Lap Time

## Set Up

In [4]:
# Import necessary data libraries.
import pandas as pd
import os 
import csv
import io
import requests
import numpy as np
import matplotlib.pyplot as plt

In [5]:
# Set up URLs.
circuits_url = 'https://raw.githubusercontent.com/georgetown-analytics/Formula1/main/data/circuits.csv'
constructor_results_url = 'https://raw.githubusercontent.com/georgetown-analytics/Formula1/main/data/constructor_results.csv'
constructor_standings_url = 'https://raw.githubusercontent.com/georgetown-analytics/Formula1/main/data/constructor_standings.csv'
constructors_url = 'https://raw.githubusercontent.com/georgetown-analytics/Formula1/main/data/constructors.csv'
driver_standings_url = 'https://raw.githubusercontent.com/georgetown-analytics/Formula1/main/data/driver_standings.csv'
drivers_url = 'https://raw.githubusercontent.com/georgetown-analytics/Formula1/main/data/drivers.csv'
lap_times_url = 'https://raw.githubusercontent.com/georgetown-analytics/Formula1/main/data/lap_times.csv'
pit_stop_url = 'https://raw.githubusercontent.com/georgetown-analytics/Formula1/main/data/pit_stops.csv'
qualifying_url = 'https://raw.githubusercontent.com/georgetown-analytics/Formula1/main/data/qualifying.csv'
races_url = 'https://raw.githubusercontent.com/georgetown-analytics/Formula1/main/data/races.csv'
results_url = 'https://raw.githubusercontent.com/georgetown-analytics/Formula1/main/data/results.csv'
seasons_url = 'https://raw.githubusercontent.com/georgetown-analytics/Formula1/main/data/seasons.csv'
status_url = 'https://raw.githubusercontent.com/georgetown-analytics/Formula1/main/data/status.csv'

In [6]:
# Set up dataframes.
circuits_df = pd.read_csv(circuits_url, sep = ',', encoding = 'latin-1')
constructor_results_df = pd.read_csv(constructor_results_url, sep = ',', engine = 'python')
constructor_standings_df = pd.read_csv(constructor_standings_url, sep = ',', engine = 'python')
constructors_df = pd.read_csv(constructor_standings_url, sep = ',', engine = 'python')
driver_standings_df = pd.read_csv(driver_standings_url, sep = ',', engine = 'python')
lap_times_df = pd.read_csv(lap_times_url, sep = ',', engine = 'python')
pit_stop_df = pd.read_csv(pit_stop_url, sep = ',', engine = 'python')
qualifying_df = pd.read_csv(constructor_standings_url, sep = ',', engine = 'python')
results_df = pd.read_csv(results_url, sep = ',', engine = 'python')
seasons_df = pd.read_csv(seasons_url, sep = ',', engine = 'python')
status_df = pd.read_csv(status_url, sep = ',', engine = 'python')
races_df = pd.read_csv(races_url, sep = ',', engine = 'c')
drivers_df = pd.read_csv(drivers_url, sep = ',', encoding = 'latin-1')

## Determine Total Laps

In [7]:
# View results_df.
results_df.head()

resultId  raceId  driverId  constructorId number  grid position  \
0         1      18         1              1     22     1        1   
1         2      18         2              2      3     5        2   
2         3      18         3              3      7     7        3   
3         4      18         4              4      5    11        4   
4         5      18         5              1     23     3        5   

  positionText  positionOrder  points  laps         time milliseconds  \
0            1              1    10.0    58  1:34:50.616      5690616   
1            2              2     8.0    58       +5.478      5696094   
2            3              3     6.0    58       +8.163      5698779   
3            4              4     5.0    58      +17.181      5707797   
4            5              5     4.0    58      +18.014      5708630   

  fastestLap rank fastestLapTime fastestLapSpeed  statusId  
0         39    2       1:27.452         218.300         1  
1         41    3       1:27.739         217.586         1  
2         41    5       1:28.090         216.719         1  
3         58    7       1:28.603         215.464         1  
4         43    1       1:27.418         218.385         1

We can see in the above `results_df` dataset that we have variables `laps` and `position`. `laps` marks the number of completed laps, `position` tells us the placement of the cars at the end of each race.

In [8]:
"""
Filter results_df so that position = 1. This only gives us the first place driver, who by taking first must
have finished the race and therefore completed the maximum number of laps.
"""
wanted_position = ['1']
position_fin = results_df[results_df['position'].isin(wanted_position)]
position_fin.head()

resultId  raceId  driverId  constructorId number  grid position  \
0          1      18         1              1     22     1        1   
22        23      19         8              6      1     2        1   
44        45      20        13              6      2     2        1   
66        67      21         8              6      1     1        1   
88        89      22        13              6      2     1        1   

   positionText  positionOrder  points  laps         time milliseconds  \
0             1              1    10.0    58  1:34:50.616      5690616   
22            1              1    10.0    56  1:31:18.555      5478555   
44            1              1    10.0    57  1:31:06.970      5466970   
66            1              1    10.0    66  1:38:19.051      5899051   
88            1              1    10.0    58  1:26:49.451      5209451   

   fastestLap rank fastestLapTime fastestLapSpeed  statusId  
0          39    2       1:27.452         218.300         1  
22         37    2       1:35.405         209.158         1  
44         38    3       1:33.600         208.153         1  
66         46    1       1:21.670         205.191         1  
88         16    3       1:26.666         221.734         1

In [9]:
# Create the variable TotalLaps. This will directly mirror the variable laps.
position_fin['TotalLaps'] = position_fin['laps']
position_fin.head()

<ipython-input-9-eb1a067850f0>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  position_fin['TotalLaps'] = position_fin['laps']


resultId  raceId  driverId  constructorId number  grid position  \
0          1      18         1              1     22     1        1   
22        23      19         8              6      1     2        1   
44        45      20        13              6      2     2        1   
66        67      21         8              6      1     1        1   
88        89      22        13              6      2     1        1   

   positionText  positionOrder  points  laps         time milliseconds  \
0             1              1    10.0    58  1:34:50.616      5690616   
22            1              1    10.0    56  1:31:18.555      5478555   
44            1              1    10.0    57  1:31:06.970      5466970   
66            1              1    10.0    66  1:38:19.051      5899051   
88            1              1    10.0    58  1:26:49.451      5209451   

   fastestLap rank fastestLapTime fastestLapSpeed  statusId  TotalLaps  
0          39    2       1:27.452         218.300         1         58  
22         37    2       1:35.405         209.158         1         56  
44         38    3       1:33.600         208.153         1         57  
66         46    1       1:21.670         205.191         1         66  
88         16    3       1:26.666         221.734         1         58

In [10]:
"""
Drop any unncessary variables. We will be keeping raceId and TotalLaps.
Note that this is in preparation for merging the data with the races_df dataset.
"""
position_fin.drop("resultId", axis = 1, inplace = True)
position_fin.drop("driverId", axis = 1, inplace = True)
position_fin.drop("constructorId", axis = 1, inplace = True)
position_fin.drop("number", axis = 1, inplace = True)
position_fin.drop("grid", axis = 1, inplace = True)
position_fin.drop("position", axis = 1, inplace = True)
position_fin.drop("positionText", axis = 1, inplace = True)
position_fin.drop("positionOrder", axis = 1, inplace = True)
position_fin.drop("points", axis = 1, inplace = True)
position_fin.drop("laps", axis = 1, inplace = True)
position_fin.drop("time", axis = 1, inplace = True)
position_fin.drop("milliseconds", axis = 1, inplace = True)
position_fin.drop("fastestLap", axis = 1, inplace = True)
position_fin.drop("rank", axis = 1, inplace = True)
position_fin.drop("fastestLapTime", axis = 1, inplace = True)
position_fin.drop("fastestLapSpeed", axis = 1, inplace = True)
position_fin.drop("statusId", axis = 1, inplace = True)

/opt/anaconda3/lib/python3.8/site-packages/pandas/core/frame.py:4167: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [11]:
# View position_fin.
position_fin.head()

raceId  TotalLaps
0       18         58
22      19         56
44      20         57
66      21         66
88      22         58

## Merge Datasets

In [12]:
# Merge position_fin with races_df by "raceId" to get final_races_df.
final_races_df = pd.merge(position_fin, races_df, on = "raceId")
final_races_df.head()

raceId  TotalLaps  year  round  circuitId                   name  \
0      18         58  2008      1          1  Australian Grand Prix   
1      19         56  2008      2          2   Malaysian Grand Prix   
2      20         57  2008      3          3     Bahrain Grand Prix   
3      21         66  2008      4          4     Spanish Grand Prix   
4      22         58  2008      5          5     Turkish Grand Prix   

         date      time                                                url  
0  2008-03-16  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...  
1  2008-03-23  07:00:00  http://en.wikipedia.org/wiki/2008_Malaysian_Gr...  
2  2008-04-06  11:30:00  http://en.wikipedia.org/wiki/2008_Bahrain_Gran...  
3  2008-04-27  12:00:00  http://en.wikipedia.org/wiki/2008_Spanish_Gran...  
4  2008-05-11  12:00:00  http://en.wikipedia.org/wiki/2008_Turkish_Gran...

In [13]:
# Merge final_races_df with results_df by raceId to get race_results_df.
race_results_df = pd.merge(final_races_df, results_df, on = "raceId")
race_results_df

raceId  TotalLaps  year  round  circuitId                   name  \
0          18         58  2008      1          1  Australian Grand Prix   
1          18         58  2008      1          1  Australian Grand Prix   
2          18         58  2008      1          1  Australian Grand Prix   
3          18         58  2008      1          1  Australian Grand Prix   
4          18         58  2008      1          1  Australian Grand Prix   
...       ...        ...   ...    ...        ...                    ...   
25199    1060         71  2021      9         70    Austrian Grand Prix   
25200    1060         71  2021      9         70    Austrian Grand Prix   
25201    1060         71  2021      9         70    Austrian Grand Prix   
25202    1060         71  2021      9         70    Austrian Grand Prix   
25203    1060         71  2021      9         70    Austrian Grand Prix   

             date    time_x  \
0      2008-03-16  04:30:00   
1      2008-03-16  04:30:00   
2      2008-03-16  04:30:00   
3      2008-03-16  04:30:00   
4      2008-03-16  04:30:00   
...           ...       ...   
25199  2021-07-04  13:00:00   
25200  2021-07-04  13:00:00   
25201  2021-07-04  13:00:00   
25202  2021-07-04  13:00:00   
25203  2021-07-04  13:00:00   

                                                     url  resultId  ...  \
0      http://en.wikipedia.org/wiki/2008_Australian_G...         1  ...   
1      http://en.wikipedia.org/wiki/2008_Australian_G...         2  ...   
2      http://en.wikipedia.org/wiki/2008_Australian_G...         3  ...   
3      http://en.wikipedia.org/wiki/2008_Australian_G...         4  ...   
4      http://en.wikipedia.org/wiki/2008_Australian_G...         5  ...   
...                                                  ...       ...  ...   
25199  http://en.wikipedia.org/wiki/2021_Austrian_Gra...     25141  ...   
25200  http://en.wikipedia.org/wiki/2021_Austrian_Gra...     25142  ...   
25201  http://en.wikipedia.org/wiki/2021_Austrian_Gra...     25143  ...   
25202  http://en.wikipedia.org/wiki/2021_Austrian_Gra...     25144  ...   
25203  http://en.wikipedia.org/wiki/2021_Austrian_Gra...     25145  ...   

       positionOrder  points laps       time_y milliseconds fastestLap  rank  \
0                  1    10.0   58  1:34:50.616      5690616         39     2   
1                  2     8.0   58       +5.478      5696094         41     3   
2                  3     6.0   58       +8.163      5698779         41     5   
3                  4     5.0   58      +17.181      5707797         58     7   
4                  5     4.0   58      +18.014      5708630         43     1   
...              ...     ...  ...          ...          ...        ...   ...   
25199             16     0.0   70           \N           \N         62    11   
25200             17     0.0   69           \N           \N         51     8   
25201             18     0.0   69           \N           \N         56    18   
25202             19     0.0   69           \N           \N         49    19   
25203             20     0.0    0           \N           \N         \N     0   

       fastestLapTime  fastestLapSpeed statusId  
0            1:27.452          218.300        1  
1            1:27.739          217.586        1  
2            1:28.090          216.719        1  
3            1:28.603          215.464        1  
4            1:27.418          218.385        1  
...               ...              ...      ...  
25199        1:08.520          226.865       11  
25200        1:08.420          227.196        4  
25201        1:09.394          224.007       12  
25202        1:09.757          222.842       12  
25203              \N               \N        4  

[25204 rows x 26 columns]

## Find Total Laps Completed by Individual Drivers

In [14]:
status_df.head()

statusId        status  familyStatus  Completion Status
0         1      Finished             4                  1
1         2  Disqualified             3                  0
2         3      Accident             1                  0
3         4     Collision             2                  0
4         5        Engine             6                  0

Note: We realized that because the "+n Laps" status still means that these drivers finished the race, they have familyStatus = 4. We can simply assign instances with familyStatus = 4 the total number of laps in a race. We do not have to bother with adding the n number of laps in their status and adding it to their current laps count. That being said, a lot of work was done with that initial idea in mind, and it is shown below. It is not work that we will use, though.

### Initial Work (not to be used)

We know from our `status_df` dataset that there are a lot of drivers with "+n Laps" as their status. Because of this, their laps count is lower than it should be. We can fix this by taking the n number of laps in their status and adding it to their current laps count.

In [15]:
# Find statuses with numbers.
status_df['Contain#'] = status_df['status'].str.contains('\\d')

# Let's take a look at the statuses with numbers.
wanted_num = [True]
status_true_num = status_df[status_df['Contain#'].isin(wanted_num)]
status_true_num

statusId     status  familyStatus  Completion Status  Contain#
10         11     +1 Lap             4                  1      True
11         12    +2 Laps             4                  1      True
12         13    +3 Laps             4                  1      True
13         14    +4 Laps             4                  1      True
14         15    +5 Laps             4                  1      True
15         16    +6 Laps             4                  1      True
16         17    +7 Laps             4                  1      True
17         18    +8 Laps             4                  1      True
18         19    +9 Laps             4                  1      True
44         45   +11 Laps             4                  1      True
49         50   +17 Laps             4                  1      True
51        128   +42 Laps             4                  1      True
52         53   +13 Laps             4                  1      True
54         55   +12 Laps             4                  1      True
57         58   +26 Laps             4                  1      True
76         77  107% Rule             3                  0      True
87         88   +10 Laps             4                  1      True
110       111   +14 Laps             4                  1      True
111       112   +15 Laps             4                  1      True
112       113   +25 Laps             4                  1      True
113       114   +18 Laps             4                  1      True
114       115   +22 Laps             4                  1      True
115       116   +16 Laps             4                  1      True
116       117   +24 Laps             4                  1      True
117       118   +29 Laps             4                  1      True
118       119   +23 Laps             4                  1      True
119       120   +21 Laps             4                  1      True
121       122   +44 Laps             4                  1      True
122       123   +30 Laps             4                  1      True
123       124   +19 Laps             4                  1      True
124       125   +46 Laps             4                  1      True
126       127   +20 Laps             4                  1      True
130       133   +49 Laps             4                  1      True
131       134   +38 Laps             4                  1      True

We can see above that all of the statuses with numbers in them are "+n Laps" statuses with one exception: "107% Rule". The statuses with "+n Laps" are 11-19, 45, 50, 128, 53, 55, 58, 88, 111-120, 122-125, 127, 133-134.

In [16]:
"""
Create a new variable, AddToLaps, that will signify the number of laps to add to a driver's initial lap count.
Do this by making a for loop through the status_true_num dataset that selects just 1:3 of the substring
in the status variable. Put these selections into a new variable, AddToLaps.

https://www.dataquest.io/blog/tutorial-advanced-for-loops-python-pandas/
https://www.guru99.com/learning-python-strings-replace-join-split-reverse.html

"""
for label, row in status_true_num.iterrows():
    status_true_num.loc[label, "AddToLaps"] = row["status"][1:3]
    
status_true_num

/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:1596: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[key] = _infer_fill_value(value)
/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:1763: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)


statusId     status  familyStatus  Completion Status  Contain# AddToLaps
10         11     +1 Lap             4                  1      True        1 
11         12    +2 Laps             4                  1      True        2 
12         13    +3 Laps             4                  1      True        3 
13         14    +4 Laps             4                  1      True        4 
14         15    +5 Laps             4                  1      True        5 
15         16    +6 Laps             4                  1      True        6 
16         17    +7 Laps             4                  1      True        7 
17         18    +8 Laps             4                  1      True        8 
18         19    +9 Laps             4                  1      True        9 
44         45   +11 Laps             4                  1      True        11
49         50   +17 Laps             4                  1      True        17
51        128   +42 Laps             4                  1      True        42
52         53   +13 Laps             4                  1      True        13
54         55   +12 Laps             4                  1      True        12
57         58   +26 Laps             4                  1      True        26
76         77  107% Rule             3                  0      True        07
87         88   +10 Laps             4                  1      True        10
110       111   +14 Laps             4                  1      True        14
111       112   +15 Laps             4                  1      True        15
112       113   +25 Laps             4                  1      True        25
113       114   +18 Laps             4                  1      True        18
114       115   +22 Laps             4                  1      True        22
115       116   +16 Laps             4                  1      True        16
116       117   +24 Laps             4                  1      True        24
117       118   +29 Laps             4                  1      True        29
118       119   +23 Laps             4                  1      True        23
119       120   +21 Laps             4                  1      True        21
121       122   +44 Laps             4                  1      True        44
122       123   +30 Laps             4                  1      True        30
123       124   +19 Laps             4                  1      True        19
124       125   +46 Laps             4                  1      True        46
126       127   +20 Laps             4                  1      True        20
130       133   +49 Laps             4                  1      True        49
131       134   +38 Laps             4                  1      True        38

In [17]:
"""
Drop the row with "107% Rule".
https://www.shanelynn.ie/pandas-drop-delete-dataframe-rows-columns/
"""
status_true_num = status_true_num.drop(labels = 76, axis = 0)
status_true_num

statusId    status  familyStatus  Completion Status  Contain# AddToLaps
10         11    +1 Lap             4                  1      True        1 
11         12   +2 Laps             4                  1      True        2 
12         13   +3 Laps             4                  1      True        3 
13         14   +4 Laps             4                  1      True        4 
14         15   +5 Laps             4                  1      True        5 
15         16   +6 Laps             4                  1      True        6 
16         17   +7 Laps             4                  1      True        7 
17         18   +8 Laps             4                  1      True        8 
18         19   +9 Laps             4                  1      True        9 
44         45  +11 Laps             4                  1      True        11
49         50  +17 Laps             4                  1      True        17
51        128  +42 Laps             4                  1      True        42
52         53  +13 Laps             4                  1      True        13
54         55  +12 Laps             4                  1      True        12
57         58  +26 Laps             4                  1      True        26
87         88  +10 Laps             4                  1      True        10
110       111  +14 Laps             4                  1      True        14
111       112  +15 Laps             4                  1      True        15
112       113  +25 Laps             4                  1      True        25
113       114  +18 Laps             4                  1      True        18
114       115  +22 Laps             4                  1      True        22
115       116  +16 Laps             4                  1      True        16
116       117  +24 Laps             4                  1      True        24
117       118  +29 Laps             4                  1      True        29
118       119  +23 Laps             4                  1      True        23
119       120  +21 Laps             4                  1      True        21
121       122  +44 Laps             4                  1      True        44
122       123  +30 Laps             4                  1      True        30
123       124  +19 Laps             4                  1      True        19
124       125  +46 Laps             4                  1      True        46
126       127  +20 Laps             4                  1      True        20
130       133  +49 Laps             4                  1      True        49
131       134  +38 Laps             4                  1      True        38

#### Merge Datasets Again

In [18]:
# Merge status_true_num with race_results_df by statusId to get race_lap_results_df.
race_lap_results_df = pd.merge(race_results_df, status_true_num, on = "statusId", how = 'outer')
race_lap_results_df.head()

raceId  TotalLaps    year  round  circuitId                   name  \
0    18.0       58.0  2008.0    1.0        1.0  Australian Grand Prix   
1    18.0       58.0  2008.0    1.0        1.0  Australian Grand Prix   
2    18.0       58.0  2008.0    1.0        1.0  Australian Grand Prix   
3    18.0       58.0  2008.0    1.0        1.0  Australian Grand Prix   
4    18.0       58.0  2008.0    1.0        1.0  Australian Grand Prix   

         date    time_x                                                url  \
0  2008-03-16  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...   
1  2008-03-16  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...   
2  2008-03-16  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...   
3  2008-03-16  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...   
4  2008-03-16  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...   

   resultId  ...  fastestLap  rank fastestLapTime  fastestLapSpeed statusId  \
0       1.0  ...          39     2       1:27.452          218.300        1   
1       2.0  ...          41     3       1:27.739          217.586        1   
2       3.0  ...          41     5       1:28.090          216.719        1   
3       4.0  ...          58     7       1:28.603          215.464        1   
4       5.0  ...          43     1       1:27.418          218.385        1   

  status  familyStatus  Completion Status  Contain# AddToLaps  
0    NaN           NaN                NaN       NaN       NaN  
1    NaN           NaN                NaN       NaN       NaN  
2    NaN           NaN                NaN       NaN       NaN  
3    NaN           NaN                NaN       NaN       NaN  
4    NaN           NaN                NaN       NaN       NaN  

[5 rows x 31 columns]

In [19]:
for label, row in race_lap_results_df.iterrows():
    if  race_lap_results_df.loc[label, "AddToLaps"] == 'NaN':
        race_lap_results_df.loc[label, "AddToLaps"].replace("NaN", 0)
    
race_lap_results_df.head()

raceId  TotalLaps    year  round  circuitId                   name  \
0    18.0       58.0  2008.0    1.0        1.0  Australian Grand Prix   
1    18.0       58.0  2008.0    1.0        1.0  Australian Grand Prix   
2    18.0       58.0  2008.0    1.0        1.0  Australian Grand Prix   
3    18.0       58.0  2008.0    1.0        1.0  Australian Grand Prix   
4    18.0       58.0  2008.0    1.0        1.0  Australian Grand Prix   

         date    time_x                                                url  \
0  2008-03-16  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...   
1  2008-03-16  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...   
2  2008-03-16  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...   
3  2008-03-16  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...   
4  2008-03-16  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...   

   resultId  ...  fastestLap  rank fastestLapTime  fastestLapSpeed statusId  \
0       1.0  ...          39     2       1:27.452          218.300        1   
1       2.0  ...          41     3       1:27.739          217.586        1   
2       3.0  ...          41     5       1:28.090          216.719        1   
3       4.0  ...          58     7       1:28.603          215.464        1   
4       5.0  ...          43     1       1:27.418          218.385        1   

  status  familyStatus  Completion Status  Contain# AddToLaps  
0    NaN           NaN                NaN       NaN       NaN  
1    NaN           NaN                NaN       NaN       NaN  
2    NaN           NaN                NaN       NaN       NaN  
3    NaN           NaN                NaN       NaN       NaN  
4    NaN           NaN                NaN       NaN       NaN  

[5 rows x 31 columns]

### Work to be Used

Like we noted earlier, the "+n Laps" status means that these drivers finished the race, and they have familyStatus = 4. We can assign instances with this familyStatus the total number of laps in a race. Instances with any other familyStatus will be assigned their completed number of laps.

#### Merge Datasets

In [20]:
# Merge race_results_df with status_df by statusId to get race_status_df.
race_status_df = pd.merge(race_results_df, status_df, on = "statusId")
race_status_df.describe()

raceId     TotalLaps          year         round     circuitId  \
count  25204.000000  25204.000000  25204.000000  25204.000000  25204.000000   
mean     516.937589     65.885177   1989.289081      8.240914     22.141287   
std      289.156434     22.676536     18.939900      4.861397     16.662924   
min        1.000000     12.000000   1950.000000      1.000000      1.000000   
25%      287.000000     54.000000   1976.000000      4.000000      9.000000   
50%      502.000000     65.000000   1990.000000      8.000000     18.000000   
75%      761.000000     75.000000   2005.000000     12.000000     32.000000   
max     1060.000000    200.000000   2021.000000     21.000000     76.000000   

           resultId      driverId  constructorId          grid  positionOrder  \
count  25204.000000  25204.000000   25204.000000  25204.000000   25204.000000   
mean   12588.639859    250.359070      47.406721     11.207507      12.933066   
std     7257.528256    257.164574      58.216417      7.273266       7.747447   
min        1.000000      1.000000       1.000000      0.000000       1.000000   
25%     6301.750000     56.000000       6.000000      5.000000       6.000000   
50%    12602.500000    159.000000      25.000000     11.000000      12.000000   
75%    18883.250000    347.000000      58.000000     17.000000      19.000000   
max    25145.000000    854.000000     214.000000     34.000000      39.000000   

             points          laps     statusId  familyStatus  \
count  25204.000000  25204.000000  25204.00000  25204.000000   
mean       1.787873     45.836613     17.76718      4.108792   
std        4.014859     30.041200     26.14447      1.377368   
min        0.000000      0.000000      1.00000      1.000000   
25%        0.000000     21.000000      1.00000      4.000000   
50%        0.000000     52.000000     11.00000      4.000000   
75%        2.000000     66.000000     15.00000      5.000000   
max       50.000000    200.000000    139.00000      6.000000   

       Completion Status  
count       25204.000000  
mean            0.547889  
std             0.497711  
min             0.000000  
25%             0.000000  
50%             1.000000  
75%             1.000000  
max             1.000000

#### Assign Laps

In [21]:
"""
Create the variable CompletedLaps.
Remember that familyStatus = 4 means that the driver finished the race and completed every lap.
"""
race_status_df['CompletedLaps'] = np.where(race_status_df.familyStatus == 4, race_status_df.TotalLaps, race_status_df.laps)
race_status_df.describe()

raceId     TotalLaps          year         round     circuitId  \
count  25204.000000  25204.000000  25204.000000  25204.000000  25204.000000   
mean     516.937589     65.885177   1989.289081      8.240914     22.141287   
std      289.156434     22.676536     18.939900      4.861397     16.662924   
min        1.000000     12.000000   1950.000000      1.000000      1.000000   
25%      287.000000     54.000000   1976.000000      4.000000      9.000000   
50%      502.000000     65.000000   1990.000000      8.000000     18.000000   
75%      761.000000     75.000000   2005.000000     12.000000     32.000000   
max     1060.000000    200.000000   2021.000000     21.000000     76.000000   

           resultId      driverId  constructorId          grid  positionOrder  \
count  25204.000000  25204.000000   25204.000000  25204.000000   25204.000000   
mean   12588.639859    250.359070      47.406721     11.207507      12.933066   
std     7257.528256    257.164574      58.216417      7.273266       7.747447   
min        1.000000      1.000000       1.000000      0.000000       1.000000   
25%     6301.750000     56.000000       6.000000      5.000000       6.000000   
50%    12602.500000    159.000000      25.000000     11.000000      12.000000   
75%    18883.250000    347.000000      58.000000     17.000000      19.000000   
max    25145.000000    854.000000     214.000000     34.000000      39.000000   

             points          laps     statusId  familyStatus  \
count  25204.000000  25204.000000  25204.00000  25204.000000   
mean       1.787873     45.836613     17.76718      4.108792   
std        4.014859     30.041200     26.14447      1.377368   
min        0.000000      0.000000      1.00000      1.000000   
25%        0.000000     21.000000      1.00000      4.000000   
50%        0.000000     52.000000     11.00000      4.000000   
75%        2.000000     66.000000     15.00000      5.000000   
max       50.000000    200.000000    139.00000      6.000000   

       Completion Status  CompletedLaps  
count       25204.000000   25204.000000  
mean            0.547889      46.485955  
std             0.497711      30.565368  
min             0.000000       0.000000  
25%             0.000000      21.000000  
50%             1.000000      53.000000  
75%             1.000000      67.000000  
max             1.000000     200.000000

#### Create a CSV file of race_status_df.

In [25]:
race_status_df.to_csv("race_status.csv", index = False)

## Find the Average Speed

The code below is currently unfinished, the syntax was giving me difficulty.

In [22]:
#race_status_df["AvgSpeed"] = int(race_status_df["milliseconds"]) / race_status_df["CompletedLaps"]
#race_status_df.head()

In [23]:
#for label, row in race_status_df.iterrows():
#    race_status_df.loc[label, "Milliseconds"] = int(row["milliseconds"])
#    race_status_df.loc[label, "AvgSpeed"] = row["milliseconds"] / row["CompletedLaps"]
    
#race_status_df.head()

In [24]:
#race_status_df = [int(milliseconds) for milliseconds in race_status_df]
#race_status_df.head()